nautobot/dcim/models/device_components.py

```diff
@@ -1096,3 +1154,101 @@ def natural_key_field_lookups(cls):  # pylint: disable=no-self-argument
         For the time being we just use the PK as a natural key.
         """
         return ["pk"]
+
+
+@extras_features(
+    "custom_links",
+    "custom_validators",
+    "export_templates",
+    "graphql",
+    "webhooks",
+)
+class ModuleBay(PrimaryModel):
+    """
+    A slot in a Device or Module which can contain Modules.
+    """
+
+    parent_device = models.ForeignKey(
+        to="dcim.Device",
+        on_delete=models.CASCADE,
+        related_name="module_bays",
+        blank=True,
+        null=True,
+    )
+    parent_module = models.ForeignKey(
+        to="dcim.Module",
+        on_delete=models.CASCADE,
+        related_name="module_bays",
+        blank=True,
+        null=True,
+    )
+    position = models.CharField(
+        blank=False,
+        null=False,
+        max_length=CHARFIELD_MAX_LENGTH,
+        help_text="The position of the module bay within the parent device/module",
+    )
+    label = models.CharField(max_length=CHARFIELD_MAX_LENGTH, blank=True, help_text="Physical label")
+    description = models.CharField(max_length=CHARFIELD_MAX_LENGTH, blank=True)
+
+    clone_fields = ["parent_device", "parent_module"]
+
+    class Meta:
+        # TODO: Ordering by parent_module.id is not correct but prevents an infinite loop
```

Comment:
Django can't resolve the multiple inheritance of `ModuleBay` -> `Module` -> `ModuleBay` -> `Module` without resulting in an infinite loop.

Reply:
Could there be alternatives like a mechanism to limit the recursion depth? We would need to understand the impact of that recursion on performance, too.

Reply:
This is in the Django SQLCompiler code. We may not want to tread there.

Reply:
Does removing `module_bay` from the Module `ordering` make it possible to order this one correctly? Might be worth considering.

Reply:
If we remove `module_bay` from module ordering then we're sorting based on location, which may be null, and then by module type. The current implementation isn't perfect but it keeps modules that belong to the same module bay parent grouped together:

```py
>>> pprint.pprint(list(Module.objects.all()))
[<Module: A10 ModuleType 3 (Asset Tag: 61083969-d305-4acc-bf90-d5216079134f) installed in rtr1 (1)>,
 <Module: Citrix ModuleType 2 (Serial: 61773776) installed in rtr1 (2)>,
 <Module: F5 ModuleType 4 (Asset Tag: b6e1afdd-0c04-48ed-a21a-7964ddbea574) installed in rtr1 (4)>,
 <Module: Palo Alto ModuleType 1 (Asset Tag: d5e77062-af1f-43c2-b0e3-c67b21d7b4d1) installed in rtr1 (5)>,
 <Module: Citrix ModuleType 2 (Serial: 18084627) installed in rtr2 (1)>,
 <Module: A10 ModuleType 3 installed in rtr2 (2)>,
 <Module: Citrix ModuleType 2 (Serial: 68045364) (Asset Tag: e600f5e8-5383-4454-9823-af9475cca1a6) installed in rtr2 (3)>,
 <Module: F5 ModuleType 4 (Serial: 78941250) installed in rtr2 (5)>]
```